storage/src/object_storage/policy.rs

```diff
@@ -0,0 +1,86 @@
+/*
+    Quickwit
+    Copyright (C) 2021 Quickwit Inc.
+
+    Quickwit is offered under the AGPL v3.0 and as commercial software.
+    For commercial licensing, contact us at hello@quickwit.io.
+
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/// The multipart policy defines when and how multipart upload / download should happen.
+///
+/// The right settings might be vendor specific, but if not available the default values
+/// should be safe.
+pub struct MultiPartPolicy {
+    /// Ideal part size.
+    /// Since S3 has a constraint on the number of parts, it cannot always be
+    /// respected.
+    pub target_part_num_bytes: usize,
+    /// Maximum number of parts allowed.
+    pub max_num_parts: usize,
+    /// Threshold above which multipart is trigged.
+    pub multipart_threshold_num_bytes: u64,
+    /// Maximum size allowed for an object.
+    pub max_object_num_bytes: u64,
+    /// Maximum number of part to be upload concurrently.
+    pub max_concurrent_upload: usize,
+}
+
+impl MultiPartPolicy {
+    /// This function returns the size of the part that should
+    /// be used. We should have `part_num_bytes(len)` <= `len`.
+    ///
+    /// If this function returns `len`, then multipart upload
+    /// will not be used.
+    pub fn part_num_bytes(&self, len: u64) -> u64 {
+        assert!(
+            len < self.max_object_num_bytes,
+            "This objet storage does not support object of that size {}",
+            self.max_object_num_bytes
+        );
+        assert!(
+            self.max_num_parts > 0,
+            "Misconfiguration: max_num_parts == 0 makes no sense."
+        );
+        if len < self.multipart_threshold_num_bytes || self.max_num_parts == 1 {
+            return len;
+        }
+        let max_num_parts = self.max_num_parts as u64;
+        // complete part is the smallest integer such that
+        // <max_num_parts> * <min_part_len> >= len.
+        let min_part_len = 1u64 + (len - 1u64) / max_num_parts;
```

Comment:
Is there anything sneaky being avoided here with the  +/- 1u64?

Reply:
Yeah we don't want a normal division. We want the definition above the formula... I don't know a simpler 
way to write it.